File: trunk/src/networking/SyncAlgorithm.java
package networking;


import java.util.LinkedList;
import java.util.List;

import networking.Client.RemoteNode;

public abstract class SyncAlgorithm {
	
	protected List<RemoteNode> network;
	protected String ip;
	protected volatile boolean isCalcDone = true;
	protected volatile boolean isPending = false;

	protected static SyncAlgorithm instance = null;

	public SyncAlgorithm(List<RemoteNode> network, String ip)
	{
		// pay attention to the difference between network and this.network
		this.network = new LinkedList<RemoteNode>();
		this.network.addAll(network);
		this.ip      = ip;
	}
	
	public synchronized boolean isCalcDone() {
		return isCalcDone;
	}
	
	public synchronized void setCalcDone(){
		isCalcDone = true;
	}
	
	public synchronized void setCalcInProgress(){
		isCalcDone = false;
	}
	
	public static SyncAlgorithm getInstance()
	{
		return instance;
	}

	abstract  public boolean canAccess();

	public synchronized void setPending() {
		System.out.println("Settig isPending");
		isPending = true;
	}
}

File: trunk/CSharp/SyncAlgorithm.cs
using System;
using System.Collections.Generic;

namespace Networking
{

    public abstract class SyncAlgorithm
    {

        protected internal List<RemoteNode> network;
        protected internal string ip;
        protected internal volatile bool isCalcDone;
        protected internal volatile bool isPending;
        protected internal volatile bool canStart;

        protected internal static SyncAlgorithm instance = null;

        public SyncAlgorithm(IList<RemoteNode> network, string ip)
        {
            canStart = false;
            isPending = false;
            isCalcDone = true;
            // pay attention to the difference between network and this.network
            this.network = new List<RemoteNode>();
            this.network.AddRange(network);
            this.ip = ip;
        }

        public virtual bool CalcDone
        {
            get
            {
                lock (this)
                {
                    return isCalcDone;
                }
            }
        }

        public virtual void setCalcDone()
        {
            lock (this)
            {
                isCalcDone = true;
            }
        }

        public virtual void setCalcInProgress()
        {
            lock (this)
            {
                isCalcDone = false;
            }
        }

        public static SyncAlgorithm Instance
        {
            get
            {
                return instance;
            }
        }

        abstract public bool canAccess();


        public virtual void setPending()

        {
            lock (this)
            {
                //Console.WriteLine("Setting isPending");
                isPending = true;
            }
        }

        public virtual void setCanStart() 
        {
            lock (this)
            {
                
             canStart = true;
            }
        }

        public bool getcanStart()
        {
            return canStart;
        }


        public void clearPending()
        {
            isPending = false;
        }

    }

}